File: Backend/Soft-Gestion-de-clientes/src/main/java/com/SoftGestionClientes/Controller/ProductController.java
package com.SoftGestionClientes.Controller;

import com.SoftGestionClientes.Dto.ProductDto;
import com.SoftGestionClientes.Exception.AlreadyRegisterException;
import com.SoftGestionClientes.Exception.BadRequestException;
import com.SoftGestionClientes.Exception.NotFoundException;
import com.SoftGestionClientes.Services.ServiceImpl.ProductServiceImpl;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;

@RestController
@RequestMapping("/pradera/products")
public class ProductController {

    @Autowired
    private ProductServiceImpl productService;

    HashMap<String, Object> data;

    /**
     * Retrieves a list of active products
     *
     * @return a response with the list of products and status OK
     */
    @GetMapping
    public ResponseEntity<Object> getAllProducts(){
        // initialize the data
        data = new HashMap<>();
        // get all products
        List<ProductDto> productsSaved = productService.getAllProducts();
        // add products to data
        data.put("data", productsSaved);
        // return a response with the data and status OK
        return new ResponseEntity<>(data, HttpStatus.OK);
    }

    /**
     * Retrieves a list of inactive products
     *
     * @return a response with the list of inactive products and status OK
     */
    @GetMapping("/inactive")
    public ResponseEntity<Object> getAllInactiveProducts(){
        // initialize the data
        data = new HashMap<>();
        // get all products inactive
        List<ProductDto> inactiveProducts = productService.getAllInactiveProducts();
        // add the products to data
        data.put("data", inactiveProducts);
        // return a response with status OK and the list with inactive products
        return new ResponseEntity<>(data, HttpStatus.OK);
    }

    /**
     * Retrieves a list of active clients filtered by a name
     * @param productName to search the clients
     * @return a response with the list of clients and status OK
     */
    @GetMapping("/{productName}")
    public ResponseEntity<Object> getProductsByName(@PathVariable String productName){
        // initialize the data
        data = new HashMap<>();
        // get the products by name
        List<ProductDto> productsSaved = productService.getProductByName(productName);
        // add products to data
        data.put("data", productsSaved);
        // return a response with status OK and products found
        return new ResponseEntity<>(data, HttpStatus.OK);
    }

    /**
     * Retrieves an active client by id
     * @param productId to search the client
     * @return a response with the client and status OK
     */
    @GetMapping("/id/{productId}")
    public ResponseEntity<Object> getProductById(@PathVariable Long productId){
        // initialize the data
        data = new HashMap<>();
        // get the product by id
        ProductDto productsSaved = productService.getProductById(productId);
        // add product to data
        data.put("data", productsSaved);
        // return a response with status OK and product found
        return new ResponseEntity<>(data, HttpStatus.OK);
    }

    /**
     * Create a product
     * @param product for to create
     * @return a response with the product created and status CREATED
     */
    @PostMapping("/register")
    public ResponseEntity<Object> registerProduct(@RequestBody ProductDto product){
        // initialize the data
        data = new HashMap<>();
        // create the new product
        ProductDto productRegistered = productService.registerProduct(product);
        // add the new product to data
        data.put("successful", productRegistered);
        // returns a response with product registered and status CREATED
        return new ResponseEntity<>(data, HttpStatus.CREATED);
    }


    /**
     * Update a product
     * @param product for to update
     * @return a response with the product updated and status OK
     */
    @PutMapping("/update")
    public ResponseEntity<Object> updatedProduct(@RequestBody ProductDto product){
        // initialize the data
        data = new HashMap<>();
        // update the product
        ProductDto productUpdated = productService.updateProduct(product);
        // add the product updated to data
        data.put("Product updated", productUpdated);
        // return a response with status OK and the product
        return new ResponseEntity<>(data, HttpStatus.OK);
    }

    /**
     * Delete a product
     * @param productId for to delete
     * @return a response with status NO CONTENT
     */
    @DeleteMapping("/{productId}")
    public ResponseEntity<Object> deleteProductById(@PathVariable Long productId){

        // delete product by id
        productService.deleteProductById(productId);
        // return a response
        return new ResponseEntity<>("Product deleted", HttpStatus.NO_CONTENT);
    }


    ////////////////////
    // EXCEPTION HANDLERS
    ////////////////////


    @ExceptionHandler(BadRequestException.class)
    public ResponseEntity<Object> handleBadReqException(BadRequestException e){
        // create a custom response to BadRequestException
        data = new HashMap<>();
        data.put("error", "Bad Request");
        data.put("message", e.getMessage());

        return new ResponseEntity<>(data, HttpStatus.BAD_REQUEST);
    }
    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<Object> handleNotFoundException(NotFoundException e){
        // create a custom response to BadRequestException
        data = new HashMap<>();
        data.put("error", "Not Found");
        data.put("message", e.getMessage());

        return new ResponseEntity<>(data, HttpStatus.NOT_FOUND);
    }
    @ExceptionHandler(AlreadyRegisterException.class)
    public ResponseEntity<Object> handleAlreadyRegisterException(AlreadyRegisterException e){
        // create a custom response to BadRequestException
        data = new HashMap<>();
        data.put("error", "Already Register");
        data.put("message", e.getMessage());

        return new ResponseEntity<>(data, HttpStatus.ALREADY_REPORTED);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<Object> handleGenericException(Exception e){
        data = new HashMap<>();
        data.put("error", "Internal Error");
        data.put("message", e.getMessage());

        return new ResponseEntity<>(data, HttpStatus.INTERNAL_SERVER_ERROR);
    }
}

File: Backend-Node/controllers/productController.ts
import { NextFunction, Request, Response } from "express"
import * as productService from '../services/productService'

/////////////////////////
// PRODUCT CONTROLLER
///////////////////////

export const registerProduct = async (req: Request, res: Response, next: NextFunction) => {
    const product = req.body // GET THE PRODUCT TO CREATE FROM THE REQUEST
    try {
        const newProduct = await productService.createProduct(product) // CREATE THE NEW PRODUCT
        res.status(201).json({ok: true, data: newProduct}) // RETURN STSTAUS 201 AND THE NEW PRODUCT    
    } catch(error){        
        next(error)
    }
}

export const updateProduct = async (req: Request, res: Response, next: NextFunction) => {
    const product = req.body // GET THE PRODUCT TO UPDATE FROM THE REQUEST
    try {
        const productUpdated = await productService.updateProduct(product) // UPDATE THE PRODUCT WITH PRODUCTSERVICE
        res.status(200).json({ok: true, data: productUpdated}) // RETURN STATUS 200 AND THE PRODUCT UPDATED
    } catch (error){
        next(error)
    }
}

export const getAllProducts = async (req: any, res: Response, next: NextFunction) => {
    const inDelivery: boolean = req.filterDelivery // CHECK IF THE FILTER IN_DELIVERY IS ACTIVE
    try {
        const products = await productService.getAllProducts(inDelivery) // GET ALL PRODUCTS AND VALIDATE WHO REQUESTS IT
        res.status(200).json({ok: true, data: products}) // RETURN STATUS 200 AND THE DATA
    } catch(error){
        next(error)
    }
}

export const getAllInactivesProducts = async (req: Request, res: Response, next: NextFunction) => {
    try {
        const inactiveProducts = await productService.getInactiveProducts()
        res.status(200).json({ok: true, data: inactiveProducts})
    } catch (error){
        next(error)
    }
}

export const getAllProductsWithName = async (req: Request, res: Response, next: NextFunction) => {
    const productName = req.params.productName // GET THE NAME FROM THE PARAMS 
    try {
        const products = await productService.getProductsByName(productName) // GET ALL PRODUCTS THAT IN HIS NAME CONTAINS PRODUCTNAME
        res.status(200).json({ok: true, data: products}) // RETURN STATUS 200 AND THE PRODUCTS
    } catch (error){
        next(error)
    }
}

export const deleteProduct = async (req: Request, res: Response, next: NextFunction) => {
    const productId = req.params.productId // GET THE PRODUCT ID FROM THE PARAMS     
    try {
        await productService.deleteProductById(productId)
        res.status(204).json({ok: true})
    } catch(error){
        next(error)
    }
}